atomic_reactor/plugins/pre_add_image_content_manifest.py

```diff
@@ -0,0 +1,186 @@
+          "dependencies": [{"purl": "pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0"}],
+          "sources": [{"purl": "pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2"}]
+        }
+      ]
+    }
+
+    WITH content_sets specified:
+
+    {
+      "metadata": {
+        "icm_version": 1,
+        "icm_spec": "https://link.to.icm.specification",
+        "image_layer_index": 2
+      },
+      "content_sets": [
+          "rhel-8-for-x86_64-baseos-rpms",
+          "rhel-8-for-x86_64-appstream-rpms"
+      ],
+      "image_contents": [
+        {
+          "purl": "pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.0.2",
+          "dependencies": [{"purl": "pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0"}],
+          "sources": [{"purl": "pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2"}]
+        }
+      ]
+    }
+    """
+    key = PLUGIN_ADD_IMAGE_CONTENT_MANIFEST
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow, icm_url, destdir=IMAGE_BUILD_INFO_DIR):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param icm_url: str, URL of the ICM from the Cachito request.
+        :param destdir: image path to carry content_manifests data dir
+        """
+        super(AddImageContentManifestPlugin, self).__init__(tasker, workflow)
+        self.content_manifests_dir = os.path.join(destdir, 'content_manifests')
+        self.icm_url = icm_url
+        self.dfp = df_parser(self.workflow.builder.df_path, workflow=self.workflow)
+        labels = Labels(self.dfp.labels)
+        _, image_name = labels.get_name_and_value(Labels.LABEL_TYPE_COMPONENT)
+        _, image_version = labels.get_name_and_value(Labels.LABEL_TYPE_VERSION)
+        _, image_release = labels.get_name_and_value(Labels.LABEL_TYPE_RELEASE)
+        self.icm_file_name = '{}-{}-{}.json'.format(image_name, image_version, image_release)
+        self.content_sets = []
+        self._layer_index = None
+        self._icm = None
+
+    @property
+    def layer_index(self):
+        if not self._layer_index:
+            # Default layer index is 1, because base and 'FROM scratch' images
+            #     *always* have 2 layers
+            self._layer_index = 1
+        if not base_image_is_scratch(self.dfp.baseimage):
+            inspect = self.workflow.builder.base_image_inspect
+            self._layer_index = len(inspect[INSPECT_ROOTFS][INSPECT_ROOTFS_LAYERS])
+        return self._layer_index
+
+    @property
+    def icm(self):
+        """
+        Get and validate the ICM from the Cachito API `content-manifest` endpoint.
+
+        :return: dict, the ICM as a Python dict
+        """
+        if not self._icm:
+            def make_session():
+                # `method_whitelist=False` allows retrying non-idempotent methods like POST
+                session = get_retrying_requests_session(method_whitelist=False)
+                session.verify = True
```

Comment:
We need to get this value from the Cachito "insecure" configuration. This will be important for development environments (such as osbs-box).

Reply:
@athos-ribeiro; I don't know what "get this value from the Cachito "insecure" configuration" means... Set `session.verify = False`? Or is there more to it?

Reply:
@ben-alkov IIUC, what athos means is to use the Cachito configuration from reactor-config-map to determine whether session should be insecure, see `get_cachito_session()` in `plugins.pre_reactor_config`.

Speaking of `get_cachito_session()`, would it make sense to add a method to download the ICM to `CachitoAPI` from `utils.cachito`?

Reply:
@chmeliik, the ICM URL is present in the initial response request, meaning it could be hosted anywhere (OSBS should not bother with it. I have no preference on whether we should download it in the plugin or in the cachito utils itself.

Reply:
@athos-ribeiro I see, in that case adding it to cachito utils probably does not make that much sense.